src/slice/sort.cr

```diff
@@ -206,4 +206,357 @@ struct Slice(T)
+  # publish, distribute, sublicense, and/or sell copies of
+  # the Software, and to permit persons to whom the Software
+  # is furnished to do so, subject to the following
+  # conditions:
+  #
+  # The above copyright notice and this permission notice
+  # shall be included in all copies or substantial portions
+  # of the Software.
+  #
+  # THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+  # ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+  # TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+  # PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+  # SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+  # CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+  # OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+  # IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+  # DEALINGS IN THE SOFTWARE.
+
+  # Slices of up to this length get sorted using insertion sort.
+  private MAX_INSERTION = 10
+
+  # Very short runs are extended using insertion sort to span at least this many elements.
+  private MIN_RUN = 10
+
+  # This merge sort borrows some (but not all) ideas from TimSort, which is described in detail
+  # [here](http://svn.python.org/projects/python/trunk/Objects/listsort.txt).
+  #
+  # The algorithm identifies strictly descending and non-descending subsequences, which are called
+  # natural runs. There is a stack of pending runs yet to be merged. Each newly found run is pushed
+  # onto the stack, and then some pairs of adjacent runs are merged until these two invariants are
+  # satisfied:
+  #
+  # 1. for every `i` in `1..runs.len()`: `runs[i - 1].len > runs[i].len`
+  # 2. for every `i` in `2..runs.len()`: `runs[i - 2].len > runs[i - 1].len + runs[i].len`
+  #
+  # The invariants ensure that the total running time is `O(n * log(n))` worst-case.
+  protected def self.merge_sort!(v : Slice(T)) forall T
+    size = v.size
+
+    # Short arrays get sorted in-place via insertion sort to avoid allocations.
+    if size <= MAX_INSERTION
+      if size >= 2
+        (size - 1).downto(0) { |i| insert_head!(v[i..]) }
+      end
+      return
+    end
+
+    # Allocate a buffer to use as scratch memory. We keep the length 0 so we can keep in it
+    # shallow copies of the contents of `v` without risking the dtors running on copies if
+    # `is_less` panics. When merging two sorted runs, this buffer holds a copy of the shorter run,
+    # which will always have length at most `len / 2`.
+    buf = Pointer(T).malloc(size // 2)
+
+    # In order to identify natural runs in `v`, we traverse it backwards. That might seem like a
+    # strange decision, but consider the fact that merges more often go in the opposite direction
+    # (forwards). According to benchmarks, merging forwards is slightly faster than merging
+    # backwards. To conclude, identifying runs by traversing backwards improves performance.
+    runs = [] of Range(Int32, Int32)
+    last = size
```

Comment:
This is particularly true here: which `size` is used, the method or the variable?